src/vsr/grid.zig

```diff
@@ -87,14 +105,17 @@ pub fn GridType(comptime Storage: type) type {
             next: ?*Read = null,
         };
 
-        pub const ReadRepair = struct {
-            callback: *const fn (*Grid.ReadRepair, error{BlockNotFound}!void) void,
-            address: u64,
-            checksum: u128,
-            block: BlockPtr,
-            grid: *Grid,
-            next_tick: Grid.NextTick = undefined,
-            completion: Storage.Read = undefined,
+        pub const ReadBlockResult = union(enum) {
+            valid: BlockPtrConst,
+            /// Checksum of block header is invalid.
+            invalid_checksum,
+            /// Checksum of block body is invalid.
+            invalid_checksum_body,
+            /// The block header is valid, but its `header.command` is not `block`.
+            /// (This is possible due to misdirected IO).
+            unexpected_command,
+            /// The block is valid, but it is not the block we expected.
+            unexpected_checksum,
```

Comment:
I am slightly uneasy here, as this errors look like something you might want to branch on, but we actually never branch IIUC, just log. Maybe

```
union(enum) {
    valid: BlockPtrConst,
    invalid: []const u8, // static string describing what's wrong
}
```

Reply:
(the risk here is that at some point in the future some code starts distinguishing between `unexpected_checksum` vs `invalid_checksum`, and that uncovereds a sleeper bug where we compute the exact variant wrong, because it only affected the logs before. Also, the precedence is not entirely obvious here, the code could get refactored to return `unexpected_checksum` before, eg,  `invalid_checsum_body`)

Reply:
What about:

```zig
union(enum) {
    valid: BlockPtrConst,
    invalid: enum {
            invalid_checksum,
            invalid_checksum_body,
            unexpected_command,
            unexpected_checksum,
    },
}
```

I'm on the fence about this.
It doesn't really fix the issues you mentioned, but maybe hints at them.

In particular, I like the `enum` for my rule-of-thumb of "don't discard information".
For example, in the future, we will have metrics, and an enum is nice since it is a static guarantee against metric label cardinality issues. (Though maybe we could assert that the string has a static lifetime to achieve something similar there...)

Reply:
Good point about metrics! :thinking: I think that maybe the current code is best overall, as it's simple and flat, and those are big advantages! Maaybe add a comment at the definition site that "although we do distinguish between the reasons why the block is invalid, we only use this info for logging".